examples/hardware_decoding/main.go

```diff
@@ -0,0 +1,195 @@
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/asticode/go-astiav"
+)
+
+var (
+	hardwareDeviceTypeName = flag.String("d", "", "the hardware device type like: cuda")
+	input                  = flag.String("i", "", "the input path")
+)
+
+type stream struct {
+	decCodec        *astiav.Codec
+	decCodecContext *astiav.CodecContext
+	hwDeviceContext *astiav.HardwareDeviceContext
+	inputStream     *astiav.Stream
+}
+
+func main() {
+	// Handle ffmpeg logs
+	astiav.SetLogLevel(astiav.LogLevelDebug)
+	astiav.SetLogCallback(func(l astiav.LogLevel, fmt, msg, parent string) {
+		log.Printf("ffmpeg log: %s (level: %d)\n", strings.TrimSpace(msg), l)
+	})
+
+	// Parse flags
+	flag.Parse()
+
+	// Usage
+	if *input == "" || *hardwareDeviceTypeName == "" {
+		log.Println("Usage: <binary path> -d <device type> -i <input path>")
+		return
+	}
+
+	// Get hardware device type
+	hardwareDeviceType := astiav.FindHardwareDeviceTypeByName(*hardwareDeviceTypeName)
+	if hardwareDeviceType == astiav.HardwareDeviceTypeNone {
+		log.Fatal(errors.New("main: hardware device not found"))
+	}
+
+	// Alloc packet
+	pkt := astiav.AllocPacket()
+	defer pkt.Free()
+
+	// Alloc hardware frame
+	hardwareFrame := astiav.AllocFrame()
+	defer hardwareFrame.Free()
+
+	// Alloc software frame
+	softwareFrame := astiav.AllocFrame()
+	defer softwareFrame.Free()
+
+	// Alloc input format context
+	inputFormatContext := astiav.AllocFormatContext()
+	if inputFormatContext == nil {
+		log.Fatal(errors.New("main: input format context is nil"))
+	}
+	defer inputFormatContext.Free()
+
+	// Open input
+	if err := inputFormatContext.OpenInput(*input, nil, nil); err != nil {
+		log.Fatal(fmt.Errorf("main: opening input failed: %w", err))
+	}
+	defer inputFormatContext.CloseInput()
+
+	// Find stream info
+	if err := inputFormatContext.FindStreamInfo(nil); err != nil {
+		log.Fatal(fmt.Errorf("main: finding stream info failed: %w", err))
+	}
+
+	// Loop through streams
+	hardwarePixelFormat := astiav.PixelFormatNone
+	streams := make(map[int]*stream) // Indexed by input stream index
+	for _, is := range inputFormatContext.Streams() {
+		var err error
+
+		// Only process video
+		if is.CodecParameters().MediaType() != astiav.MediaTypeVideo {
+			continue
+		}
+
+		// Create stream
+		s := &stream{inputStream: is}
+
+		// Find decoder
+		if s.decCodec = astiav.FindDecoder(is.CodecParameters().CodecID()); s.decCodec == nil {
+			log.Fatal(errors.New("main: codec is nil"))
+		}
+
+		// Alloc codec context
+		if s.decCodecContext = astiav.AllocCodecContext(s.decCodec); s.decCodecContext == nil {
+			log.Fatal(errors.New("main: codec context is nil"))
+		}
+		defer s.decCodecContext.Free()
+
+		codec_hardware_configs, err := s.decCodec.HardwareConfigs(hardwareDeviceType)
```

Comment:
Could you
- add a comment `// Get codec hardware configs`
- rename `codec_hardware_configs` to `hardwareConfigs`
- remove the error handling since `.HardwareConfigs` now only returns the slice